fvm/src/kernel/default.rs

```diff
@@ -864,6 +876,114 @@ where
             .create_actor(code_id, actor_id, delegated_address)
     }
 
+    fn can_actor_upgrade(&self) -> bool {
+        let mut iter = self.call_manager.get_actor_call_stack().iter();
+
+        // find the first position of this actor on the call stack
+        let position = iter.position(|&tuple| tuple == (self.actor_id, INVOKE_FUNC_NAME));
+        if position.is_none() {
+            return true;
+        }
+
+        // make sure that no other actor appears on the call stack after 'position' (unless its
+        // a recursive upgrade call which is allowed)
+        for tuple in iter {
+            if tuple.0 != self.actor_id || tuple.1 != UPGRADE_FUNC_NAME {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    fn upgrade_actor<K: Kernel>(
+        &mut self,
+        new_code_cid: Cid,
+        params_id: BlockId,
+    ) -> Result<SendResult> {
+        if self.read_only {
+            return Err(
+                syscall_error!(ReadOnly, "upgrade_actor cannot be called while read-only").into(),
+            );
+        }
+
+        // Load parameters.
+        let params = if params_id == NO_DATA_BLOCK_ID {
+            None
+        } else {
+            Some(self.blocks.get(params_id)?.clone())
+        };
+
+        // Make sure we can actually store the return block.
+        if self.blocks.is_full() {
+            return Err(syscall_error!(LimitExceeded; "cannot store return block").into());
+        }
+
+        let result = self.call_manager.with_transaction(|cm| {
+            let origin = cm.origin();
+
+            let state = cm
+                .get_actor(self.actor_id)?
+                .ok_or_else(|| syscall_error!(NotFound; "actor not found"))?;
```

Comment:
I think the correct error here is `IllegalOperation`. This can happen if we, e.g., try to upgrade after selfdestructing (which _could_ happen, in theory).

Reply:
Cool yes `IllegalOperation` definetly makes sense here